### ISSUE TITLE
coroutine was never awaited

### ISSUE THREAD

**Issue by andrea-cassioli-maersk** [Azure/azure-sdk-for-python#21639]
- **Operating System**: linux 
- **Python Version**: 3.8.8

**Describe the bug**

On termination we get `sys:1: RuntimeWarning: coroutine 'Lock.acquire' was never awaited` error message. 

Given there is no `acquire` in our code and we only create tasks, I am assuming some of the async packages are doing so.

**To Reproduce**
I do not have a clear way to reproduce.

**Expected behavior**
Clean exit, no error messages.

**Additional context**
Our code handles graceful termination, i.e. we process SIGTERM to ensure our app shuts down nicely. That means when SIGTERM is received a specific exception is thrown and handled specifically. 

My guess is that this exception breaks some library flow, leaving a not still awaited coroutine pending. 

Maybe similar to https://github.com/Azure/azure-sdk-for-python/issues/21061

**Comment by xiangyan99** [comment]
Thanks for reporting the issue. Could you help to give more information?

e.g. which libraries are you using? Which version? And if possible, could you give us the repro steps?

Thank you.

**Comment by andrea-cassioli-maersk** [comment]
@xiangyan99 here the libraries:

azure-common==1.1.27
azure-servicebus==7.3.3
azure-storage-blob==12.9.0

both storage and service bus are used in an async fashion.

It is quite difficult to reproduce without an informed guess on which library might cause it. Searching the sdk code base I have found a number of network/authentication related pieces of code calling `acquire` but there might be other. My guess is that in some cases an async call is not awaited but rather a coroutine is created.

**Comment by xiangyan99** [comment]
Thank you. Do you know the azure-core version?

Because https://github.com/Azure/azure-sdk-for-python/issues/21061 is fixed in azure-core 1.20.0. Just want to check if it is the same issue or not.

**Comment by andrea-cassioli-maersk** [comment]
Yes, it is azure-core==1.18.0

**Comment by xiangyan99** [comment]
Thank you.

Is there any chance you can try azure-core 1.20.0?

**Comment by andrea-cassioli-maersk** [comment]
I might, but it does not seem to be the same coroutine to me

**Comment by xiangyan99** [comment]
Thank you for the information. We will investigate ASAP.

**Comment by xiangyan99** [comment]
Are you using communication libraries? I do see some communication libraries call Lock.acquire() w/o await.

Not sure if they are the ones caused your problem.

**Comment by andrea-cassioli-maersk** [comment]
indeed we also use aiohttp, but I cannot find any mention of `Lock.acquire` in their code base. But I will post on their github as well tomorrow.

**Comment by iscai-msft** [comment]
@xiangyan99 this shouldn't be related to azure.core / azure.core.rest as far as I know, I don't think we have any `Lock.acquire` code there. Is there any possibility of getting a stacktrace, or does the code handling in place get rid of this information?

**Comment by xiangyan99** [comment]
@andrea-cassioli-maersk I meant azure-communication-identity,  azure-communication-networktraversal, etc.

Are you using these libraries?

Sorry for the confusion.

**Comment by andrea-cassioli-maersk** [comment]
@iscai-msft we only get that message when the application exits. Not sure I can increase the verbosity.

@xiangyan99 not directly at least. Here all the "azure" packages

```
azure-cli-core==2.28.1
azure-cli-telemetry==1.0.6
azure-common==1.1.27
azure-core==1.18.0
azure-keyvault==4.1.0
azure-keyvault-certificates==4.3.0
azure-keyvault-keys==4.4.0
azure-keyvault-secrets==4.3.0
azure-mgmt-core==1.2.2
azure-servicebus==7.3.3
azure-storage-blob==12.9.0
```

**Comment by iscai-msft** [comment]
talked with @annatisch and she thinks it's most likely either `azure-servicebus` or `azure-storage-blob`. @andrea-cassioli-maersk are you doing any blob uploads or message lock renewals in service bus?

cc @yunhaoling for servicebus just in case

**Comment by andrea-cassioli-maersk** [comment]
@iscai-msft as a matter of fact we do both!

**Comment by annatisch** [comment]
Thanks @andrea-cassioli-maersk - how are you renewing the service bus message locks?
For example - are you using the auto-lock-renew feature?

**Comment by andrea-cassioli-maersk** [comment]
@annatisch we use `AutoLockRenewer` to keep lock on messages

**Comment by annatisch** [comment]
Thanks @andrea-cassioli-maersk - is it possible that the code is using the synchronous AutoLockRenewer for messages received from an async client?
The async client will need to be paired with an async AutoLockRenewer - as seen here:
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/samples/async_samples/auto_lock_renew_async.py

**Comment by andrea-cassioli-maersk** [comment]
@annatisch we are using all async code. It looks a bit like
```
msg_batch = await receiver.receive_messages(max_message_count=1, max_wait_time=queue_read_timeout)
        if msg_batch:
            max_lock_duration = 20 * 60

            try:
                renewer = AutoLockRenewer(max_lock_renewal_duration=max_lock_duration, on_lock_renew_failure=on_lock_renew_failure_callback)

                msg = msg_batch[0]
                renewer.register(receiver, msg, max_lock_renewal_duration=max_lock_duration)

                await callback(msg, receiver)
            finally:
                await renewer.close()
```

**Comment by annatisch** [comment]
Thanks @andrea-cassioli-maersk 
Adding @yunhaoling from the messaging crew to take a closer look :)

I will double check the Storage SDK as well to see if there's any locks being used during upload that may cause this.

**Comment by yunhaoling** [comment]
as far as I know we are not using async lock in the Service Bus Auto Lock Renewer or in the Service Bus code.

However, async Lock is used in the uamqp, e.g., [ascyn connection lock](https://github.com/Azure/azure-uamqp-python/blob/master/uamqp/async_ops/connection_async.py#L111)

So I was wondering if it's the async connection that got interrupted before existing it gracefully.
hey @andrea-cassioli-maersk , do you use `ServiceBusReceiver` with context manager or shutdown the receiver in your shutdowning process like the following:

```python
async with receiver:
    # your code

# or

await receiver.close()
```
also looping in @swathipil to help further investigate into the issue from the service bus side

**Comment by andrea-cassioli-maersk** [comment]
@yunhaoling we use the receiver as a context manager

**Comment by swathipil** [comment]
@andrea-cassioli-maersk 
Would you be able to share your logs if you have any? 
If not, would you be able to enable uamqp logging for ServiceBus (by following [this section](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/servicebus/azure-servicebus#logging)) and Storage Blobs (by following [this section](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-blob/README.md#logging)), then share?

These may give us a clue about which (if either) of the two SDKs that @iscai-msft suspected are causing this issue.

**Comment by andrea-cassioli-maersk** [comment]
@swathipil I cannot really add more logging (especially uamqp that is very verbose...). We have not seen this issue happening for an entire week, so I guess it must be something happening in a very corner case. I have add some more log in our code to see if I can pin down a bit better where that might happen in our code.

**Comment by andrea-cassioli-maersk** [comment]
Hey, I got this new example (with a bit of better logging)

```
2021-11-22T23:04:40.707Z | RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2021-11-22T23:04:40.707Z | await asyncio.wait_for(self._async_lock.acquire(), timeout=timeout, **self._internal_kwargs)
2021-11-22T23:04:40.707Z | /usr/local/lib/python3.8/site-packages/uamqp/async_ops/connection_async.py:111: RuntimeWarning: coroutine 'Lock.acquire' was never awaited
2021-11-22T23:04:40.707Z | RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2021-11-22T23:04:40.707Z | await asyncio.wait_for(self._async_lock.acquire(), timeout=timeout, **self._internal_kwargs)
2021-11-22T23:04:40.707Z | /usr/local/lib/python3.8/site-packages/uamqp/async_ops/connection_async.py:111: RuntimeWarning: coroutine 'wait_for' was never awaited
```

It definitely seems to be related to uamqp.

**Comment by swathipil** [comment]
@andrea-cassioli-maersk These are super helpful, thank you!

This is definitely under uamqp, so we'll have a fix out for this asap!

**Comment by andrea-cassioli-maersk** [comment]
@swathipil you are welcome! Is there any timeline for the patch release?

**Comment by swathipil** [comment]
@andrea-cassioli-maersk - So I took a look at where the issue might be happening (according to the logs), and it looks like the call to `asyncio.wait_for(self._async_lock.acquire()...)` seems to be correct. More specifically, according to the [`asyncio` docs, `wait_for`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for) does take a coroutine as the first parameter. In this case, the coroutine `Lock.acquire()` is being passed correctly to `wait_for`, so this warning shouldn't be popping up.

The logs are helpful in pinpointing where the RuntimeWarning is happening, but unfortunately, I'm not sure _why_ yet. I'm continuing to dig into this and appreciate your patience!

A couple more questions for you:
1) You mentioned that there is a specific exception thrown. Do you have the specific exception? (Do we know whether this exception is caused by the RuntimeWarning?)
2) Do you have a snippet of code that can stably reproduce this error?

**Comment by andrea-cassioli-maersk** [comment]
Hi @swathipil,

1. well our code handles graceful termination, i.e. we process SIGTERM to ensure our app shuts down nicely. That means when SIGTERM is received a custom exception is thrown from the signal handler. 
2. Not really
3. Not really, I posted what I have. My understanding, based on very basic examples, is that the runtimewarning is thrown when the code exits.

**Comment by andrea-cassioli-maersk** [comment]
Btw, I have by chance noticed that in the newly released python 3.10 there is a fix for this bug: https://bugs.python.org/issue37658

**Comment by andrea-cassioli-maersk** [comment]
Any thoughts?

**Comment by yunhaoling** [comment]
hey @andrea-cassioli-maersk , apologize for the late response.

I'm not sure if the python bug is related to the issue you meet into here. My understanding is that the python bug is more about not able to free socket (leaks) when the async open connection coroutine task is cancelled.

However, here the issue we ran into is that the async coroutine task (wait_for) does not get executed when we interrupt the application to stop.

I want to reproduce the issue first, so I spent some time to play with the signal module in python and used following scripts for experiment:

```python
# This is a mock application

import asyncio
import signal
import os


class SignalHandler:
    def __init__(self):
        self.exit = False
        signal.signal(signal.SIGTERM, self.handler)
        signal.signal(signal.SIGINT, self.handler)

    def handler(self, *args):
        print('dealing with signal', *args)
        self.exit = True


async def worker(lock, exit_handler):
    while not exit_handler.exit:
        await asyncio.wait_for(lock.acquire(), timeout=None)
        await asyncio.sleep(1)
        print('working')
        lock.release()


async def main(exit_handler):
    lock = asyncio.Lock()
    res = [asyncio.create_task(worker(lock, exit_handler)) for _ in range(3)]

    cnt = 0
    while not handler.exit:  # main coroutine, do nothing
        await asyncio.sleep(1)
        cnt += 1

    print('await all the tasks to finish')
    for future in res:
        await future  # wait for all the coroutine tasks to finish

print('Current pid', os.getpid())
handler = SignalHandler()
asyncio.run(main(handler))
```

I start this scrip first, then I try the following two approach to stop the program:
- CTRL + C: the handler is triggered with SIGINT, and all the tasks are awaited to be finished.
<img width="318" alt="image" src="https://user-images.githubusercontent.com/47871814/146287475-2f247bef-d9c0-4786-895d-c6c80778efc0.png">

- Kill the program via a python command os.kill, the handler doesn't get executed and program finishes without tasks being awaited
<img width="183" alt="image" src="https://user-images.githubusercontent.com/47871814/146287254-0a59c7b6-24a4-42a5-85f1-7d68e09deb03.png">

<img width="93" alt="image" src="https://user-images.githubusercontent.com/47871814/146287312-36138377-608f-44aa-a921-dfb2f3b39280.png">


I have little clue on what's happening here and need to understand more what python is doing...

- Are you doing something similar to mine which cleans up the resources? if so, could you confirm that those clean up steps are executed?
- Is this issue blocking your application, or it's more of that the warning logging is annoyance?

**Comment by andrea-cassioli-maersk** [comment]
Hi @yunhaoling , sorry for the late reply!

Well, the main difference with our setup is that in the signal handler we raise an exception like 

```
    def handler(self, *args):
        print('dealing with signal', *args)
        raise GracefulTerminationException
```
the worker then handled the exception and exit.

As for your last question, even thought the logging is indeed annoying (especially because we have alerts based on log severity), we would like to make sure this is not the warning sign of any issue.

**Comment by andrea-cassioli-maersk** [comment]
Hey Team, any update?

**Comment by swathipil** [comment]
Hey @andrea-cassioli-maersk - sorry for the late response + really appreciate your patience! We've been pulled aside by a few other things but will definitely be digging into this again in this upcoming week.
Thanks!

**Comment by yunhaoling** [comment]
hey @andrea-cassioli-maersk , I have been playing with the asyncio lib recently, and I feel it could be not handling asyncio error properly under the situation of "asyncio.wait_for" + interrupt.

I'm still working on getting some simplified code to reproduce the issue first.

**Comment by yunhaoling** [comment]
hey @andrea-cassioli-maersk , apologize for the late reply.

I tried multiple code snippets but didn't get good luck. what I'm thinking right now is that based on the logging:

```python
2021-11-22T23:04:40.707Z | RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2021-11-22T23:04:40.707Z | await asyncio.wait_for(self._async_lock.acquire(), timeout=timeout, **self._internal_kwargs)
2021-11-22T23:04:40.707Z | /usr/local/lib/python3.8/site-packages/uamqp/async_ops/connection_async.py:111: RuntimeWarning: coroutine 'wait_for' was never awaited
```

- Is there a chance that the async event loop got closed when the program is interrupted while the coroutines are already added to the async event loop event queue but don't get a chance to be executed -- when cleaning up the async event loop queue, those coroutines cannot be executed with a closing/closed event loop leading to `was never awaiter`.
- have you seen other errors related asyncio event loop like event loop closed?
- is your receiving wrapped in a separate coroutine/future task and did you "await" it to finish when exiting the program?

here's the code snippet I have in mind for a graceful coroutine task shutdown:


```python

async def revc_work():
    receiver = ...
    async with receiver:
        while True:
            msg_batch = await receiver.receive_messages(max_message_count=1, max_wait_time=queue_read_timeout)
            if msg_batch:
                max_lock_duration = 20 * 60
                try:
                    renewer = AutoLockRenewer(max_lock_renewal_duration=max_lock_duration, on_lock_renew_failure=on_lock_renew_failure_callback)
    
                    msg = msg_batch[0]
                    renewer.register(receiver, msg, max_lock_renewal_duration=max_lock_duration)
    
                    await callback(msg, receiver)
                finally:
                    await renewer.close()

async def main():
try:
    recv_task = asyncio.create_task(revc_work())
    asyncio.sleep(10000000000)  # basically run forever until asked to stop
except:
    await recv_task
```

**Comment by andrea-cassioli-maersk** [comment]
hey @yunhaoling ,
sorry for the late reply.  Regarding to your three points:

* yes, this is what I consider the most likely scenario
* no, we have seen no other asyncio related errors so far
* the receiving is in the main thread and it looks a bit like

```
@contextmanager
def handle_worker_termination_signal():
    """
    Set a custom handler that throws a custom exception when termination signal SIGTERM is received.
    Once out of scope, the original handler gets restored
    """
    original_handler = signal.signal(signal.SIGTERM, _handler_termination)
    try:
        yield
    finally:
        signal.signal(signal.SIGTERM, original_handler)

async def main():
 with handle_worker_termination_signal():
     while True:
        try:
            await receive_and_process()
        except GracefulTerminationException:
            break

# main entry point of application
if __name__ == '__main__':
    asyncio.run(main())
```

**Comment by swathipil** [comment]
Hi @andrea-cassioli-maersk - We just released [azure-servicebus 7.10.0](https://pypi.org/project/azure-servicebus/), which uses a pure Python based AMQP implementation rather than `uamqp`. Because this warning originates from the `uamqp` library, this should no longer be an issue. If you find any issues with our newly released version, please open another issue in this repo. Thanks! 🙂
